sql/postgresql/pggrammar.g4

```diff
@@ -0,0 +1,4294 @@
+         | qualified_name PERCENT TYPE_P
+;
+ opt_array_bounds: opt_array_bounds OPEN_BRACKET CLOSE_BRACKET
+                 | opt_array_bounds OPEN_BRACKET iconst CLOSE_BRACKET
+                 |
+;
+ simpletypename: generictype
+               | numeric
+               | bit
+               | character
+               | constdatetime
+               | constinterval opt_interval
+               | constinterval OPEN_PAREN iconst CLOSE_PAREN
+;
+ consttypename: numeric
+              | constbit
+              | constcharacter
+              | constdatetime
+;
+ generictype: type_function_name opt_type_modifiers
+            | type_function_name attrs opt_type_modifiers
+;
+ opt_type_modifiers: OPEN_PAREN expr_list CLOSE_PAREN
+                   |
+;
+ numeric: INT_P
+        | INTEGER
+        | SMALLINT
+        | BIGINT
+        | REAL
+        | FLOAT_P opt_float
+        | DOUBLE_P PRECISION
+        | DECIMAL_P opt_type_modifiers
+        | DEC opt_type_modifiers
+        | NUMERIC opt_type_modifiers
+        | BOOLEAN_P
+;
+ opt_float: OPEN_PAREN iconst CLOSE_PAREN
+          |
+;
+ bit: bitwithlength
+    | bitwithoutlength
+;
+ constbit: bitwithlength
+         | bitwithoutlength
+;
+ bitwithlength: BIT opt_varying OPEN_PAREN expr_list CLOSE_PAREN;
+
+ bitwithoutlength: BIT opt_varying;
+
+character: characterWithLength
+          | characterWithoutLength
+;
+ constcharacter: characterWithLength
+               | characterWithoutLength
+;
+ characterWithLength: character_c OPEN_PAREN iconst CLOSE_PAREN;
+
+ characterWithoutLength: character_c;
+
+ character_c: CHARACTER opt_varying
+          | CHAR_P opt_varying
+          | VARCHAR
+          | NATIONAL CHARACTER opt_varying
+          | NATIONAL CHAR_P opt_varying
+          | NCHAR opt_varying
+;
+ opt_varying: VARYING
+            |
+;
+ constdatetime: TIMESTAMP OPEN_PAREN iconst CLOSE_PAREN opt_timezone
+              | TIMESTAMP opt_timezone
+              | TIME OPEN_PAREN iconst CLOSE_PAREN opt_timezone
+              | TIME opt_timezone
+;
+ constinterval: INTERVAL;
+//TODO with_la was used
+ opt_timezone: WITH TIME ZONE
+             | WITHOUT TIME ZONE
+             |
+;
+ opt_interval: YEAR_P
+             | MONTH_P
+             | DAY_P
+             | HOUR_P
+             | MINUTE_P
+             | interval_second
+             | YEAR_P TO MONTH_P
+             | DAY_P TO HOUR_P
+             | DAY_P TO MINUTE_P
+             | DAY_P TO interval_second
+             | HOUR_P TO MINUTE_P
+             | HOUR_P TO interval_second
+             | MINUTE_P TO interval_second
+             |
+;
+ interval_second: SECOND_P
+                | SECOND_P OPEN_PAREN iconst CLOSE_PAREN
+;
+ a_expr: c_expr
+       | a_expr TYPECAST typename
+       | a_expr COLLATE any_name
+       | a_expr AT TIME ZONE a_expr
+       | PLUS a_expr
+       | MINUS a_expr
+
+       | a_expr STAR a_expr
+       | a_expr SLASH a_expr
+       | a_expr PERCENT a_expr
```

Comment:
Not sure it's correct. [The documentation](https://www.postgresql.org/docs/7.2/sql-precedence.html) says these operators have the same precedence, so they should be merged into one alternative. Also, please take a look at how [left recursive](https://github.com/antlr/antlr4/blob/master/doc/left-recursion.md) rules work.

```suggestion
       | a_expr (STAR | SLASH | PERCENT) a_expr
```

It's actual for other operators too.

Reply:
the same consideration as above - keep grammar closer to original, even if it may be optimized.
currently, diffs are only to respect precedences and to avoid mutual left recursion problems in simple_select rule

Reply:
It's not about optimization but about correctness.

```sql
a / a * a
```

For the old grammar this input is parsed as follows (`STAR` has higher priority than `SLASH`):

```
 div
 /  \
a    *
    /  \
   a    a
```

Is it correct? I think the valid output is (`STAR` has the `SLASH` priority):

```
       *
     /   \
   div    a
 /     \
a       a
```

Could you point to your reference documentation?

Reply:
fixed precedences according to https://www.postgresql.org/docs/12/sql-syntax-lexical.html#SQL-PRECEDENCE
@KvanTTT